### PR TITLE
Add actions to create, execute, get monitors api

### PR DESCRIPTION
**Pull request by skkosuri-amzn**
*Issue #, if available:*
#232
*Description of changes:*
This to add actions to create/modify , execute, get monitor.

#232
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: alerting/src/test/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/ExecuteMonitorRequestTests.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    fun `test execute monitor request with monitor`() {
+        val monitor = randomMonitor().copy(inputs = listOf(SearchInput(emptyList(), SearchSourceBuilder())))
+        val req = ExecuteMonitorRequest(false, TimeValue.timeValueSeconds(100L), null, monitor)
+        assertNotNull(req.monitor)
+
+        val out = BytesStreamOutput()
+        req.writeTo(out)
+        val sin = StreamInput.wrap(out.bytes().toBytesRef().bytes)
+        val newReq = ExecuteMonitorRequest(sin)
+        assertNull(newReq.monitorId)
```
Comment (lezzago) [inline alerting/src/test/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/ExecuteMonitorRequestTests.kt:51]: Please compare this with the original monitorId

File: alerting/src/test/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/GetMonitorResponseTests.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        val testInstance = Instant.ofEpochSecond(1538164858L)
+
+        val cronSchedule = CronSchedule(cronExpression, ZoneId.of("Asia/Kolkata"), testInstance)
+        val req = GetMonitorResponse("1234", 1L, 2L, 0L, RestStatus.OK,
+                Monitor(
+                "123",
+                        0L, "test-monitor", true, cronSchedule, Instant.now(),
+                        Instant.now(), 0, mutableListOf(), mutableListOf(), mutableMapOf()
+
+                ))
```
Comment (lezzago) [inline alerting/src/test/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/GetMonitorResponseTests.kt:54]: Please fix formatting so it is consistent for the input attributes

File: alerting/src/test/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/IndexMonitorResponseTests.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        val testInstance = Instant.ofEpochSecond(1538164858L)
+
+        val cronSchedule = CronSchedule(cronExpression, ZoneId.of("Asia/Kolkata"), testInstance)
+        val req = IndexMonitorResponse("1234", 1L, 2L, 0L, RestStatus.OK,
+                Monitor(
+                "123",
+                        0L, "test-monitor", true, cronSchedule, Instant.now(),
+                        Instant.now(), 0, mutableListOf(), mutableListOf(), mutableMapOf()
+
+                ))
```
Comment (lezzago) [inline alerting/src/test/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/IndexMonitorResponseTests.kt:40]: Please fix formatting so it is consistent for the input attributes

File: alerting/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/GetMonitorResponse.kt [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    @Throws(IOException::class)
+    constructor(sin: StreamInput) : this(
+        sin.readString(), // id
+        sin.readLong(), // version
+        sin.readLong(), // seqNo
+        sin.readLong(), // primaryTerm
+        sin.readEnum(RestStatus::class.java), // RestStatus
+        if (sin.readBoolean()) {
+            Monitor.readFrom(sin) // monitor
+        } else null
```
[the remark this answers is not in the document]
Comment (skkosuri-amzn) [reply alerting/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/GetMonitorResponse.kt:65]: This response is used for GET or HEAD request. With HEAD request we wont have monitor object in response. This checks if monitor is present.  Also refer to writeTo
      ```if (monitor != null) {
            out.writeBoolean(true)
            monitor?.writeTo(out)
        } else {
            out.writeBoolean(false)
        }```

File: alerting/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/transport/TransportExecuteMonitorAction.kt [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        if (execMonitorRequest.monitorId != null) {
+            val getRequest = GetRequest(ScheduledJob.SCHEDULED_JOBS_INDEX).id(execMonitorRequest.monitorId)
+            client.get(getRequest, object : ActionListener<GetResponse> {
+                override fun onResponse(response: GetResponse) {
+                    if (!response.isExists) {
+                        actionListener.onFailure(
+                            ElasticsearchStatusException("Can't find monitor with id: ${response.id}", RestStatus.NOT_FOUND)
+                        )
+                    }
+                    if (!response.isSourceEmpty) {
```
[the remark this answers is not in the document]
Comment (skkosuri-amzn) [reply alerting/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/transport/TransportExecuteMonitorAction.kt:73]: Good catch. I add this extra check here to be extra safe. We were doing that in RestGetMonitorAction https://github.com/opendistro-for-elasticsearch/alerting/blob/9f405c22f5f5e1e37c442935b3235e03cacf0762/alerting/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/resthandler/RestGetMonitorAction.kt#L91

File: alerting/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/transport/TransportIndexMonitorAction.kt [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    @Volatile private var requestTimeout = REQUEST_TIMEOUT.get(settings)
+    @Volatile private var indexTimeout = INDEX_TIMEOUT.get(settings)
+    @Volatile private var maxActionThrottle = MAX_ACTION_THROTTLE_VALUE.get(settings)
+
+    init {
+        clusterService.clusterSettings.addSettingsUpdateConsumer(ALERTING_MAX_MONITORS) { maxMonitors = it }
+        clusterService.clusterSettings.addSettingsUpdateConsumer(REQUEST_TIMEOUT) { requestTimeout = it }
+        clusterService.clusterSettings.addSettingsUpdateConsumer(INDEX_TIMEOUT) { indexTimeout = it }
+        clusterService.clusterSettings.addSettingsUpdateConsumer(MAX_ACTION_THROTTLE_VALUE) { maxActionThrottle = it }
+    }
```
[the remark this answers is not in the document]
Comment (skkosuri-amzn) [reply alerting/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/transport/TransportIndexMonitorAction.kt:87]: Good catch. I missed to remove this from RestIndexMonitorAction. Will fix it.

File: alerting/src/test/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/ExecuteMonitorRequestTests.kt (hunk 1 above)
Comment (skkosuri-amzn) [reply alerting/src/test/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/ExecuteMonitorRequestTests.kt:51]: In this test the monitorId is null, so left this to check `assertNull`. Added a new check below to compare monitor.

File: alerting/src/test/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/IndexMonitorResponseTests.kt (hunk 3 above)
Comment (skkosuri-amzn) [reply alerting/src/test/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/IndexMonitorResponseTests.kt:40]: Fixed formatting.

File: alerting/src/test/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/GetMonitorResponseTests.kt (hunk 2 above)
Comment (skkosuri-amzn) [reply alerting/src/test/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/GetMonitorResponseTests.kt:54]: Fixed formatting.

File: alerting/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/GetMonitorResponse.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (lezzago) [reply alerting/src/main/kotlin/com/amazon/opendistroforelasticsearch/alerting/action/GetMonitorResponse.kt:65]: Misread the code, just realized that this is inside a constructor call. Makes sense why this is done

**Pull request merged**